net/libfab/libfab.c

```diff
@@ -940,33 +958,36 @@ static void libfab_txep_comp_read(struct fid_cq *cq, struct m0_fab__tm *tm)
  */
 static void libfab_poller(struct m0_fab__tm *tm)
 {
+	struct m0_net_end_point  *net;
 	struct m0_fab__ev_ctx    *ctx;
 	struct m0_fab__ep        *xep;
 	struct m0_fab__active_ep *aep;
 	struct fid_cq            *cq;
 	struct epoll_event        ev;
 	int                       ev_cnt;
 	int                       ret;
+	int                       err;
 
 	libfab_tm_event_post(tm, M0_NET_TM_STARTED);
 	while (tm->ftm_state != FAB_TM_SHUTDOWN) {
+		ret = fi_trywait(tm->ftm_fab->fab_fab, tm->ftm_fids.ftf_head,
+				 tm->ftm_fids.ftf_cnt);
 		/*
-		 * The lab team observed a significant increase in the CPU load
-		 * due to the epoll_wait not sleeping.
-		 * The proposed short term fix to support the lab team is adding
-		 * nanosleep of 0ns on VM to reduce CPU load.
-		 */
-		if (m0_processor_is_vm())
-			m0_nanosleep(M0_MKTIME(0 ,0), NULL);
-		/*
-		 * It is observed that with epoll_wait,
-		 * the thread is waiting in a busy-loop for events
-		 * thus not releasing CPU.
-		 * Hence, adding a sched_yield() will release the CPU for
-		 * other processes and reduce CPU consumption.
+		 * TBD : Add handling of other return values of fi_trywait() if
+		 * it returns something other than -EAGAIN and 0.
 		 */
-		sched_yield();
-		ev_cnt = epoll_wait(tm->ftm_epfd, &ev, 1, FAB_WAIT_FD_TMOUT);
+		M0_ASSERT(M0_IN(ret, (0, -EAGAIN)));
+		if (ret == 0) {
+			err = -EINTR;
+			while (err == -EINTR) {
+				ret = epoll_wait(tm->ftm_epfd, &ev, 1,
+						 FAB_WAIT_FD_TMOUT);
+				M0_ASSERT(M0_IN(ret, (-1, 0, 1)));
+				err = ret < 0 ? errno : 0;
```

Comment:
1. Let's use `-errno` here (instead of just `errno`).
2. `errno` could be overwritten somewhere in M0_ASSERT. Let's make sure we do not end up with this problem by saving errno right after the epoll_wait call.
3. Let's assert that -EINTR is the only error that we may get from epoll_wait.

I guess it would look like that:

```c
ret = epoll_wait(...);
/*
    M0_ERR is omitted because we expect only one particular error,
    and this error gets handled by the loop.
*/
err = ret < 0 ? -errno : 0;
M0_ASSERT_INFO(M0_IN(ret, (-1, 0, 1)), "Unexpected return code from epoll_wait: %d", ret);
M0_ASSERT_INFO(ergo(ret < 0, err == -EINTR), "Unexpected epoll_wait error: %d", err);
```

Reply:
Updated.